File: docs/srs.js
// SM-2 Spaced Repetition – port of srs.py

const MASTERED_REPS = 3;
const UNLOCK_THRESHOLD = 0.9;

function cardId(chapterNum, section, polish) {
  return `${chapterNum}:${section}:${polish}`;
}

function reverseCardId(chapterNum, section, polish) {
  return `rev:${chapterNum}:${section}:${polish}`;
}

// Compute a normalised base key for grouping gender/form pairs.
// Strips parentheticals first ("a Pole (male)" → "a pole"), then strips
// gender words at word boundaries so suffix pairs also group:
//   "an Englishman" → "an english",  "an Englishwoman" → "an english"
// "Germany" is safe: "many" is not a word-boundary match for "man".
function genderNormBase(english) {
  // 1. Strip parentheticals: "a Pole (male)" → "a Pole"
  // 2. Lowercase
  // 3. Strip standalone male/female: "a German (male)" → "a german"
  // 4. Strip man/woman compound suffixes: "Englishman" → "english"
  //    Uses \w{4,} so "german" (prefix "ger" = 3 chars) is NOT stripped.
  return english
    .replace(/\s*\([^)]*\)/g, '')
    .trim()
    .toLowerCase()
    .replace(/\b(male|female)\b/gi, '')
    .replace(/\b(\w{4,})woman\b/gi, '$1')
    .replace(/\b(\w{4,})man\b/gi, '$1')
    .replace(/\s+/g, ' ')
    .trim();
}

// Build a map of genderNormBase → [english strings] for one chapter's vocabulary.
function buildGenderGroups(chapterData) {
  const groups = {};
  for (const sec of chapterData.vocabulary) {
    for (const item of sec.items) {
      const base = genderNormBase(item.english);
      if (!groups[base]) groups[base] = [];
      groups[base].push(item.english);
    }
  }
  return groups;
}

function loadSrs() {
  try {
    const raw = localStorage.getItem('srs_state');
    return raw ? JSON.parse(raw) : {};
  } catch (_) { return {}; }
}

function saveSrs(state) {
  localStorage.setItem('srs_state', JSON.stringify(state));
}

function todayStr() {
  return new Date().toISOString().split('T')[0];
}

function addDays(n) {
  const d = new Date();
  d.setDate(d.getDate() + n);
  return d.toISOString().split('T')[0];
}

function daysSince(dateStr) {
  if (!dateStr) return null;
  return Math.round((new Date(todayStr()) - new Date(dateStr)) / 86400000);
}

function updateCard(state, cid, correct) {
  const card = state[cid] || { interval: 1, ease: 2.5, reps: 0, lapses: 0 };
  const quality = correct ? 4 : 1;

  if (quality < 3) {
    card.interval = 1;
    card.lapses = (card.lapses || 0) + 1;
    card.reps = 0;
    delete card.mastered_on;  // lapsed — reset mastered date
  } else {
    if (card.reps === 0) {
      card.interval = 1;
    } else if (card.reps === 1) {
      card.interval = 6;
    } else {
      card.interval = Math.round(card.interval * card.ease);
    }
    card.reps += 1;
  }

  card.ease = Math.max(1.3, card.ease + 0.1 - (5 - quality) * 0.08);
  card.due = addDays(card.interval);

  // Attempt counters
  card.total = (card.total || 0) + 1;
  if (correct) card.correct = (card.correct || 0) + 1;

  // Record when card first reaches (or re-reaches after lapse) mastered threshold
  if (correct && card.reps >= MASTERED_REPS && !card.mastered_on) {
    card.mastered_on = todayStr();
    card.mastered_in = card.total;  // total attempts at time of first mastery
  }

  state[cid] = card;
  return state;
}

function chapterMastery(srsState, chapterData) {
  let total = 0, mastered = 0;
  for (const sec of chapterData.vocabulary) {
    for (const item of sec.items) {
      const fwd = cardId(chapterData.chapter, sec.section, item.polish);
      const rev = reverseCardId(chapterData.chapter, sec.section, item.polish);
      total += 2;
      if ((srsState[fwd]?.reps || 0) >= MASTERED_REPS) mastered++;
      if ((srsState[rev]?.reps || 0) >= MASTERED_REPS) mastered++;
    }
  }
  return total ? mastered / total : 0;
}

function chapterUnlockInfo(allChapters, chaptersData) {
  if (!allChapters.length) {
    return { active: [], frontier: null, mastery: 0, nextNum: null, masteries: {} };
  }
  const srsState = loadSrs();
  const active = [allChapters[0]];
  const masteries = {};

  for (let i = 0; i < allChapters.length - 1; i++) {
    const n = allChapters[i];
    const m = chapterMastery(srsState, chaptersData[n]);
    masteries[n] = m;
    if (m >= UNLOCK_THRESHOLD) {
      active.push(allChapters[i + 1]);
    } else {
      break;
    }
  }

  const frontier = active[active.length - 1];
  if (!(frontier in masteries)) {
    masteries[frontier] = chapterMastery(srsState, chaptersData[frontier]);
  }

  const idx = allChapters.indexOf(frontier);
  const nextNum = idx + 1 < allChapters.length ? allChapters[idx + 1] : null;

  return { active, frontier, mastery: masteries[frontier], nextNum, masteries };
}

function makeAnswerChars(answer) {
  const letterIndices = [];
  for (let i = 0; i < answer.length; i++) {
    if (/\p{L}/u.test(answer[i])) letterIndices.push(i);
  }
  if (!letterIndices.length) {
    return answer.split('').map(c => ({ char: c, gap: false }));
  }
  const nGaps = Math.max(1, Math.min(letterIndices.length - 1, Math.round(letterIndices.length * 0.4)));
  const shuffled = [...letterIndices].sort(() => Math.random() - 0.5);
  const gapSet = new Set(shuffled.slice(0, nGaps));
  return answer.split('').map((c, i) => ({ char: c, gap: gapSet.has(i) }));
}

function getStats(allChapters, chaptersData) {
  const srsState = loadSrs();
  let total = 0, notStarted = 0, learning = 0, mastered = 0;

  const forecastMap = {};
  for (let i = 0; i < 7; i++) forecastMap[addDays(i)] = 0;

  function countCard(card) {
    if (!card) return 'notStarted';
    if (card.reps >= MASTERED_REPS) return 'mastered';
    return 'learning';
  }

  const chapterStats = [];
  for (const n of allChapters) {
    const ch = chaptersData[n];
    let chTotal = 0, chNotStarted = 0, chLearning = 0, chMastered = 0;
    for (const sec of ch.vocabulary) {
      for (const item of sec.items) {
        // Count forward and reverse cards separately
        for (const cid of [cardId(n, sec.section, item.polish), reverseCardId(n, sec.section, item.polish)]) {
          const card = srsState[cid];
          const bucket = countCard(card);
          total++; chTotal++;
          if (bucket === 'notStarted') { notStarted++; chNotStarted++; }
          else if (bucket === 'mastered') { mastered++; chMastered++; if (card.due in forecastMap) forecastMap[card.due]++; }
          else { learning++; chLearning++; if (card.due in forecastMap) forecastMap[card.due]++; }
        }
      }
    }
    chapterStats.push({ number: n, topic: ch.topic, total: chTotal, notStarted: chNotStarted, learning: chLearning, mastered: chMastered });
  }

  const forecast = Object.keys(forecastMap).sort().map((date, i) => ({
    date,
    count: forecastMap[date],
    label: new Date(date + 'T12:00:00').toLocaleDateString('en', { weekday: 'short' }),
    isToday: i === 0,
  }));
  const maxForecast = Math.max(...forecast.map(f => f.count), 1);

  return { total, notStarted, learning, mastered, forecast, maxForecast, chapters: chapterStats };
}

function getChapterWordStats(chapterNum, chapterData) {
  const srsState = loadSrs();
  const items = [];
  for (const sec of chapterData.vocabulary) {
    for (const item of sec.items) {
      const fwd = srsState[cardId(chapterNum, sec.section, item.polish)] || {};
      const rev = srsState[reverseCardId(chapterNum, sec.section, item.polish)] || {};
      items.push({
        english: item.english,
        polish: item.polish,
        section: sec.section,
        fwd: { total: fwd.total||0, correct: fwd.correct||0, reps: fwd.reps||0, mastered_on: fwd.mastered_on||null, mastered_in: fwd.mastered_in||null },
        rev: { total: rev.total||0, correct: rev.correct||0, reps: rev.reps||0, mastered_on: rev.mastered_on||null, mastered_in: rev.mastered_in||null },
      });
    }
  }
  return items;
}

function getDueCards(activeChapters, chaptersData, extraDays = 0) {
  const srsState = loadSrs();
  const cutoff = addDays(extraDays);
  const due = [];
  for (const n of activeChapters) {
    const ch = chaptersData[n];
    const genderGroups = buildGenderGroups(ch);
    for (const sec of ch.vocabulary) {
      for (const item of sec.items) {
        const fwdId = cardId(n, sec.section, item.polish);
        const revId = reverseCardId(n, sec.section, item.polish);

        if (!srsState[fwdId] || srsState[fwdId].due <= cutoff) {
          due.push({
            type: 'flashcard',
            prompt: item.english,
            answer: item.polish,
            hint: item.pronunciation,
            section: sec.section,
            cardId: fwdId,
          });
        }
        if (!srsState[revId] || srsState[revId].due <= cutoff) {
          const group = genderGroups[genderNormBase(item.english)];
          const revCard = {
            type: 'flashcard_reverse',
            prompt: item.polish,
            answer: item.english,
            hint: item.pronunciation,
            section: sec.section,
            cardId: revId,
          };
          if (group && group.length > 1) revCard.choices = group;
          due.push(revCard);
        }
      }
    }
  }
  return due;
}

File: srs.py
"""SM-2 spaced repetition for Polish vocabulary."""

import json
import random
from datetime import date, timedelta
from pathlib import Path

DATA_DIR = Path(__file__).parent / "data"
SRS_FILE = DATA_DIR / "srs.json"


def _today() -> str:
    return date.today().isoformat()


def load_srs() -> dict:
    if SRS_FILE.exists():
        with open(SRS_FILE, encoding="utf-8") as f:
            return json.load(f)
    return {}


def save_srs(state: dict) -> None:
    with open(SRS_FILE, "w", encoding="utf-8") as f:
        json.dump(state, f, ensure_ascii=False, indent=2)


def card_id(chapter_num: int, section: str, polish: str) -> str:
    return f"{chapter_num}:{section}:{polish}"


def update_card(state: dict, cid: str, correct: bool) -> dict:
    """Apply SM-2 update to a card. Returns updated state dict."""
    card = state.get(cid, {"interval": 1, "ease": 2.5, "reps": 0, "lapses": 0})

    quality = 4 if correct else 1

    if quality < 3:
        # Wrong: reset
        card["interval"] = 1
        card["lapses"] = card.get("lapses", 0) + 1
        card["reps"] = 0
    else:
        reps = card["reps"]
        if reps == 0:
            card["interval"] = 1
        elif reps == 1:
            card["interval"] = 6
        else:
            card["interval"] = round(card["interval"] * card["ease"])
        card["reps"] += 1

    # Adjust ease factor
    card["ease"] = max(1.3, card["ease"] + 0.1 - (5 - quality) * 0.08)

    due = date.today() + timedelta(days=card["interval"])
    card["due"] = due.isoformat()

    state[cid] = card
    return state


def _load_chapter(n: int) -> dict:
    with open(DATA_DIR / f"chapter{n}.json", encoding="utf-8") as f:
        return json.load(f)


MASTERED_REPS = 3      # consecutive correct answers to be considered mastered
UNLOCK_THRESHOLD = 0.9  # chapter mastery required to unlock the next chapter


def chapter_mastery(state: dict, chapter_num: int) -> float:
    """Return fraction of vocab cards in chapter_num that are mastered."""
    ch = _load_chapter(chapter_num)
    total = mastered = 0
    for sec in ch["vocabulary"]:
        for item in sec["items"]:
            cid = card_id(chapter_num, sec["section"], item["polish"])
            total += 1
            card = state.get(cid)
            if card and card.get("reps", 0) >= MASTERED_REPS:
                mastered += 1
    return mastered / total if total else 0.0


def chapter_unlock_info(all_chapters: list[int]) -> dict:
    """Return unlock state for all chapters in a single SRS state load.

    Returns:
        active      – ordered list of unlocked chapter numbers
        frontier    – highest active chapter number
        mastery     – mastery fraction of the frontier chapter
        next_num    – next chapter number to unlock, or None
        masteries   – dict mapping chapter_num -> mastery fraction (active only)
    """
    if not all_chapters:
        return {"active": [], "frontier": None, "mastery": 0.0, "next_num": None, "masteries": {}}

    state = load_srs()
    active = [all_chapters[0]]
    masteries: dict[int, float] = {}

    for i, n in enumerate(all_chapters[:-1]):
        m = chapter_mastery(state, n)
        masteries[n] = m
        if m >= UNLOCK_THRESHOLD:
            active.append(all_chapters[i + 1])
        else:
            break

    frontier = active[-1]
    if frontier not in masteries:
        masteries[frontier] = chapter_mastery(state, frontier)

    idx = all_chapters.index(frontier)
    next_num = all_chapters[idx + 1] if idx + 1 < len(all_chapters) else None

    return {
        "active": active,
        "frontier": frontier,
        "mastery": masteries[frontier],
        "next_num": next_num,
        "masteries": masteries,
    }


def make_answer_chars(answer: str) -> list[dict]:
    """Return a list of {char, gap} dicts for gapped-word rendering.

    ~40% of alphabetic characters are randomly chosen as gaps.  Spaces and
    punctuation are always shown.  At least one letter is always visible.
    """
    letter_indices = [i for i, c in enumerate(answer) if c.isalpha()]
    if not letter_indices:
        return [{"char": c, "gap": False} for c in answer]
    n_gaps = max(1, min(len(letter_indices) - 1, round(len(letter_indices) * 0.4)))
    gap_set = set(random.sample(letter_indices, n_gaps))
    return [{"char": c, "gap": i in gap_set} for i, c in enumerate(answer)]


def get_stats(chapters: list[int]) -> dict:
    """Return progress statistics across all chapters."""
    state = load_srs()
    today = date.today()

    total = not_started = learning = mastered = 0
    forecast = {(today + timedelta(days=i)).isoformat(): 0 for i in range(7)}
    chapter_stats = []

    for n in chapters:
        ch = _load_chapter(n)
        ch_total = ch_not_started = ch_learning = ch_mastered = 0

        for sec in ch["vocabulary"]:
            for item in sec["items"]:
                cid = card_id(n, sec["section"], item["polish"])
                card = state.get(cid)
                total += 1
                ch_total += 1

                if card is None:
                    not_started += 1
                    ch_not_started += 1
                elif card.get("reps", 0) >= MASTERED_REPS:
                    mastered += 1
                    ch_mastered += 1
                    if card.get("due") in forecast:
                        forecast[card["due"]] += 1
                else:
                    learning += 1
                    ch_learning += 1
                    if card.get("due") in forecast:
                        forecast[card["due"]] += 1

        chapter_stats.append({
            "number": n,
            "topic": ch["topic"],
            "total": ch_total,
            "not_started": ch_not_started,
            "learning": ch_learning,
            "mastered": ch_mastered,
        })

    forecast_list = [
        {"date": d, "count": forecast[d], "label": date.fromisoformat(d).strftime("%a")}
        for d in sorted(forecast)
    ]
    max_forecast = max((f["count"] for f in forecast_list), default=1) or 1

    return {
        "total": total,
        "not_started": not_started,
        "learning": learning,
        "mastered": mastered,
        "forecast": forecast_list,
        "max_forecast": max_forecast,
        "chapters": chapter_stats,
    }


def get_due_cards(chapters: list[int], extra_days: int = 0) -> list[dict]:
    """Return list of vocab items due for review.

    extra_days > 0 extends the window beyond today, enabling study-ahead.
    Cards with no SRS record (never seen) are always included.
    """
    state = load_srs()
    cutoff = (date.today() + timedelta(days=extra_days)).isoformat()
    due = []

    for n in chapters:
        ch = _load_chapter(n)
        for sec in ch["vocabulary"]:
            for item in sec["items"]:
                cid = card_id(n, sec["section"], item["polish"])
                card = state.get(cid)
                if card is None or card["due"] <= cutoff:
                    due.append({
                        "type": "flashcard",
                        "prompt": item["english"],
                        "answer": item["polish"],
                        "hint": item["pronunciation"],
                        "section": sec["section"],
                        "card_id": cid,
                    })

    return due
